### PR TITLE
Upgrade kyuubi-hive-jdbc hive version to 3.1.3

### PR DESCRIPTION
**Pull request by gabrywu**
### Why are the changes needed?

upgrade hive to 3.1.3 in the kyuubi-hive-jdbc module

### change details

file|comment
--- | ---
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/HttpBasicAuthInterceptor.java | add customCookies as a construct parameter
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/HttpKerberosRequestInterceptor.java | replace assumeSubject field with `Subject loggedInSubject` ; add customCookies as a construct parameter
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/HttpRequestInterceptorBase.java | add customCookies as a construct parameter; add customCookies initialization codes
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/HttpTokenAuthInterceptor.java | add customCookies as a construct parameter
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/JdbcColumn.java | support TIMESTAMP_WITH_TIMEZONE & UNION_TYPE type
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiBaseResultSet.java | support TIMESTAMPLOCALTZ_TYPE
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiConnection.java | add loggedInSubject,wmPool & wmApp fields and their initialization codes; add getAllUrls function; use Utils.getCanonicalHostName to extract host from connParams; clear hive conf in embedded mode; use SessionUtils.getTokenStrForm to get token string; remove isZkDynamicDiscoveryMode
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiDatabaseMetaData.java | support getSQLKeywords; add Object class type parameter to KyuubiMetaDataResultSet
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiMetaDataResultSet.java | optimize initialization logic of data & columnTypes field; add normalizeColumnNames function
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiStatement.java | split closeClientOperation into 2 functions(closeStatementIfNeeded & closeClientOperation) & add reInitState function to replace initFlags & ad error messages to exceptions in CANCELED_STATE
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/Utils.java | add dummyAuthorityString, SERVICE_DISCOVERY_MODE_ZOOKEEPER_HA, ZOOKEEPER_ACTIVE_PASSIVE_HA_DEFAULT_NAMESPACE, WM_POOL, HTTP_COOKIE_PREFIX, APPLICATION fields; replace authorityList field with suppliedURLAuthority in JdbcConnectionParams;add a construct function to JdbcConnectionParams; replace getAuthorityList function with getSuppliedURLAuthority; split parseURL into 2 different functions(parseURL, extractURLComponents); add configureConnParamsFromZooKeeper, handleAllDeprecations, getAuthorityFromJdbcURL functions; remove configureConnParams;remove joinStringArray;add getCanonicalHostName function
kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/ZooKeeperHiveClientHelper.java | add more functions
kyuubi-server/src/main/resources/META-INF/services/org.apache.hadoop.security.token.TokenIdentifier | change package name of DelegationTokenIdentifier
kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HiveDelegationTokenProvider.scala | change package name of DelegationTokenIdentifier
kyuubi-server/src/test/scala/org/apache/kyuubi/credentials/HiveDelegationTokenProviderSuite.scala | add configs, METASTORE_CONNECTION_POOLING_TYPE = none,METASTORE_AUTO_CREATE_ALL = true, METASTORE_SCHEMA_VERIFICATION = false; add construct parameter clientConf to HadoopThriftAuthBridgeWithServerContextClassLoader
kyuubi-common/src/test/scala/org/apache/kyuubi/operation/SparkMetadataTests.scala | remove getSQLKeywords from unsupported list
externals/kyuubi-hive-sql-engine/pom.xml | add HikariCP to test scope
pom.xml | update hive version to 3.1.3
kyuubi-hive-jdbc-shaded/pom.xml | add hive-standalone-metastore dependency; add hive-llap-client dependency
kyuubi-hive-jdbc/pom.xml | add hive-standalone-metastore dependency;add hive-llap-client dependency
kyuubi-server/pom.xml | add hive-standalone-metastore dependency
dev/dependencyList | add hive-standalone-metastore
kyuubi-hive-jdbc-shaded/src/main/resources/META-INF/LICENSE | update content because of hive-standalone-metastore & hive-llap-client
kyuubi-hive-jdbc-shaded/src/main/resources/META-INF/NOTICE | update content because of hive-standalone-metastore & hive-llap-client

### How was this patch tested?

use existing UTs to test this patch

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -183,7 +183,7 @@
         <maven.plugin.javadoc.version>3.3.0</maven.plugin.javadoc.version>
         <!-- DO NOT bump 4.4.0, see https://github.com/apache/incubator-kyuubi/pull/441 -->
         <!-- DO NOT bump 4.5.3, see https://github.com/apache/incubator-kyuubi/issues/708 -->
-        <maven.plugin.scala.version>4.3.0</maven.plugin.scala.version>
+        <maven.plugin.scala.version>4.5.6</maven.plugin.scala.version>
```
Comment (pan3793) [inline pom.xml:186]: Unnecessary change?

Comment (gabrywu) [reply pom.xml:186]: I will rollback it and try again. It's to resolve some compiling errors

**Comment by gabrywu** [comment]
@pan3793 @turboFei  Please don't review this PR until all of the checks are successful

File: pom.xml (hunk 1 above)
Comment (gabrywu) [reply pom.xml:186]: I tried, but failed to compile projects. It is necessary

File: externals/kyuubi-hive-sql-engine/pom.xml [hunk 2]
```diff
@@ -159,63 +165,63 @@
...
+                <configuration>
+                    <shadedArtifactAttached>false</shadedArtifactAttached>
+                    <artifactSet>
                         <includes>
-                            <include>org.apache.curator.**</include>
+                            <include>org.scala-lang:scala-library</include>
+                            <include>com.fasterxml.jackson.core:jackson-annotations</include>
+                            <include>com.fasterxml.jackson.core:jackson-core</include>
+                            <include>com.fasterxml.jackson.core:jackson-databind</include>
+                            <include>com.fasterxml.jackson.module:jackson-module-scala_2.12</include>
```
Comment (pan3793) [inline externals/kyuubi-hive-sql-engine/pom.xml:179]: Not related, but seems we include jackson w/o relocating, it may cause potential risk for class confliction.

I don't see content diff, it's just a format change? If yes, please revert this part to minimize the diff of this PR, we can handle it in seperated PR.

File: externals/kyuubi-hive-sql-engine/pom.xml [hunk 3]
```diff
@@ -147,7 +147,13 @@
             <version>${project.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.zaxxer</groupId>
+            <artifactId>HikariCP</artifactId>
+            <version>2.6.1</version>
```
Comment (pan3793) [inline externals/kyuubi-hive-sql-engine/pom.xml:153]: Extract it as a maven property and manage it under `<dependencyManagement>` block in root pom before using.

File: externals/kyuubi-hive-sql-engine/pom.xml [hunk 4]
```diff
@@ -147,7 +147,13 @@
...
+            <scope>test</scope>
 
+        </dependency>
```
Comment (pan3793) [inline externals/kyuubi-hive-sql-engine/pom.xml:156]: nit: switch line 155 and 156

File: externals/kyuubi-hive-sql-engine/pom.xml (hunk 2 above)
Comment (gabrywu) [reply externals/kyuubi-hive-sql-engine/pom.xml:179]: A separate PR is better

File: kyuubi-hive-jdbc-shaded/pom.xml [hunk 5]
```diff
@@ -112,6 +112,19 @@
             </exclusions>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-standalone-metastore</artifactId>
```
Comment (pan3793) [inline kyuubi-hive-jdbc-shaded/pom.xml:117]: r u missing `hive-llap-client`?

Comment (pan3793) [reply kyuubi-hive-jdbc-shaded/pom.xml:117]: we should udpate LICENSE and NOTICE file under the shaded module when change the shaded content.

File: externals/kyuubi-hive-sql-engine/pom.xml (hunk 3 above)
Comment (gabrywu) [reply externals/kyuubi-hive-sql-engine/pom.xml]: done

File: kyuubi-hive-jdbc-shaded/pom.xml (hunk 5 above)
Comment (gabrywu) [reply kyuubi-hive-jdbc-shaded/pom.xml:117]: > r u missing `hive-llap-client`?

hive-llap-client is in kyuubi-hive-jdbc/pom.xml

Comment (gabrywu) [reply kyuubi-hive-jdbc-shaded/pom.xml:117]: > we should udpate LICENSE and NOTICE file under the shaded module when change the shaded content.

done

**Comment by gabrywu** [comment]
https://github.com/apache/incubator-kyuubi/issues/2492

File: kyuubi-server/pom.xml [hunk 6]
```diff
@@ -245,10 +257,6 @@
                     <groupId>org.apache.hadoop</groupId>
                     <artifactId>*</artifactId>
                 </exclusion>
-                <exclusion>
-                    <groupId>org.apache.hive</groupId>
-                    <artifactId>*</artifactId>
```
Comment (pan3793) [inline kyuubi-server/pom.xml:250]: We manage hive dependency manually instead of using maven transitive dependencies, because the hive dependencies is likely out of control if we do that. Please restore this exclusion and add everything explicitly.

PS: this rule only apply to hive, in most cases, we prefer use transitive dependencies.

File: pom.xml [hunk 7]
```diff
@@ -202,6 +202,7 @@
 
         <!-- Package to use when relocating shaded classes. -->
         <kyuubi.shade.packageName>org.apache.kyuubi.shade</kyuubi.shade.packageName>
+        <hikaricp.version>2.6.1</hikaricp.version>
```
Comment (pan3793) [inline pom.xml:205]: move to line 119

File: kyuubi-hive-jdbc-shaded/src/main/resources/META-INF/NOTICE [hunk 8]
```diff
@@ -41,22 +41,28 @@ Curator Framework
...
-Copyright 2021 The Apache Software Foundation
+Copyright 2022 The Apache Software Foundation
 
 Hive Service
-Copyright 2021 The Apache Software Foundation
+Copyright 2022 The Apache Software Foundation
 
 Hive Service RPC
-Copyright 2019 The Apache Software Foundation
+Copyright 2022 The Apache Software Foundation
```
Comment (pan3793) [inline kyuubi-hive-jdbc-shaded/src/main/resources/META-INF/NOTICE:65]: Should we also update Hive Shims 0.23 and Hive Shims Common's year?

**Comment by pan3793** [comment]
Thanks @gabrywu for working on this, I think it is most there, @turboFei would you please check if this upgrade break anything you added?

File: externals/kyuubi-hive-sql-engine/pom.xml (hunk 2 above)
Comment (gabrywu) [reply externals/kyuubi-hive-sql-engine/pom.xml:179]: rollback this change

File: kyuubi-server/pom.xml (hunk 6 above)
Comment (gabrywu) [reply kyuubi-server/pom.xml]: done

File: pom.xml (hunk 7 above)
Comment (gabrywu) [reply pom.xml]: done

File: kyuubi-hive-jdbc-shaded/src/main/resources/META-INF/NOTICE (hunk 8 above)
Comment (gabrywu) [reply kyuubi-hive-jdbc-shaded/src/main/resources/META-INF/NOTICE:65]: no, their version not updated

**Review by pan3793 — APPROVED** [review]

File: pom.xml [hunk 9]
```diff
@@ -1339,6 +1340,13 @@
                 <artifactId>gson</artifactId>
                 <version>${gson.version}</version>
             </dependency>
+
+            <dependency>
+                <groupId>com.zaxxer</groupId>
+                <artifactId>HikariCP</artifactId>
+                <version>${hikaricp.version}</version>
```
Comment (turboFei) [inline pom.xml:1347]: how about adding `test` scope here?

Comment (pan3793) [reply pom.xml:1347]: We don't manage `<scope>` in `<dependencyManagement>` block

**Comment by turboFei** [comment]
thanks, merged to master

**Comment by gabrywu** [comment]
> thanks, merged to master

it's perfect
